File: frontend/types/document.ts
export type DocumentType = "receipt" | "notice" | "document" | "memo" | "other";
export type ProcessingStatus = "uploaded" | "queued" | "processing" | "ready" | "needs_review" | "confirmed" | "completed" | "failed";

export interface FolderSummary {
  label: string;
  value: string;
  count: number;
  needs_review: number;
  confirmed: number;
  processing: number;
}

export interface DocumentRecord {
  id: string;
  original_filename: string;
  stored_file_path: string;
  mime_type: string;
  source_file_type: string | null;
  extraction_method: string | null;
  ingestion_metadata: Record<string, unknown> | null;
  document_type: DocumentType;
  title: string | null;
  raw_text: string | null;
  extracted_date: string | null;
  extracted_amount: string | null;
  subtotal: string | null;
  tax: string | null;
  currency: string | null;
  merchant_name: string | null;
  category: string | null;
  tags: string[];
  summary: string | null;
  confidence_score: string | null;
  ai_document_type: DocumentType | null;
  ai_confidence_score: string | null;
  ai_extraction_notes: string | null;
  review_required: boolean;
  extraction_provider: string | null;
  refinement_provider: string | null;
  provider_chain: string | null;
  merge_strategy: string | null;
  field_sources: Record<string, string> | null;
  workflow_summary: string | null;
  action_items: string[];
  warnings: string[];
  key_dates: string[];
  urgency_level: "low" | "medium" | "high" | null;
  follow_up_required: boolean;
  workflow_metadata: Record<string, unknown> | null;
  is_favorite: boolean;
  processing_status: ProcessingStatus;
  preview_image_path: string | null;
  processing_error: string | null;
  created_at: string;
  updated_at: string;
  file_url: string;
}

export interface DocumentListResponse {
  items: DocumentRecord[];
  total: number;
  page: number;
  page_size: number;
}

export interface DocumentStats {
  total: number;
  receipts: number;
  notices: number;
  completed: number;
  processing: number;
  failed: number;
  needs_review: number;
  queued: number;
  confirmed: number;
  recent: DocumentRecord[];
  recent_updated: DocumentRecord[];
  recent_review: DocumentRecord[];
  pinned: DocumentRecord[];
  category_overview: FolderSummary[];
  file_type_overview: FolderSummary[];
}

export type DocumentUpdate = Pick<
  DocumentRecord,
  | "document_type"
  | "title"
  | "raw_text"
  | "extracted_date"
  | "extracted_amount"
  | "subtotal"
  | "tax"
  | "currency"
  | "merchant_name"
  | "category"
  | "tags"
  | "summary"
> & {
  confidence_score?: string | null;
  processing_status?: ProcessingStatus;
  is_favorite?: boolean;
};

export interface ActivitySummary {
  recent_uploads: DocumentRecord[];
  recent_edits: DocumentRecord[];
  recent_needs_review: DocumentRecord[];
  favorites: DocumentRecord[];
}

File: backend/app/schemas/document.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.document import DocumentType, ProcessingStatus


class DocumentBase(BaseModel):
    document_type: DocumentType = DocumentType.other
    title: str | None = Field(default=None, max_length=255)
    raw_text: str | None = None
    extracted_date: date | None = None
    extracted_amount: Decimal | None = Field(default=None, ge=0)
    subtotal: Decimal | None = Field(default=None, ge=0)
    tax: Decimal | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, max_length=8)
    merchant_name: str | None = Field(default=None, max_length=255)
    category: str | None = Field(default=None, max_length=80)
    tags: list[str] = Field(default_factory=list)
    summary: str | None = None


class DocumentUpdate(DocumentBase):
    confidence_score: Decimal | None = Field(default=None, ge=0, le=1)


class DocumentRead(DocumentBase):
    id: UUID
    original_filename: str
    stored_file_path: str
    mime_type: str
    source_file_type: str | None = None
    extraction_method: str | None = None
    ingestion_metadata: dict | None = None
    confidence_score: Decimal | None = None
    ai_document_type: DocumentType | None = None
    ai_confidence_score: Decimal | None = None
    ai_extraction_notes: str | None = None
    review_required: bool = False
    extraction_provider: str | None = None
    refinement_provider: str | None = None
    provider_chain: str | None = None
    merge_strategy: str | None = None
    field_sources: dict[str, str] | None = None
    workflow_summary: str | None = None
    action_items: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    key_dates: list[str] = Field(default_factory=list)
    urgency_level: str | None = None
    follow_up_required: bool = False
    workflow_metadata: dict | None = None
    processing_status: ProcessingStatus
    preview_image_path: str | None = None
    processing_error: str | None = None
    created_at: datetime
    updated_at: datetime
    file_url: str

    model_config = ConfigDict(from_attributes=True)


class DocumentListResponse(BaseModel):
    items: list[DocumentRead]
    total: int
    page: int
    page_size: int


class DocumentStats(BaseModel):
    total: int
    receipts: int
    notices: int
    completed: int
    processing: int
    failed: int
    needs_review: int = 0
    queued: int = 0
    recent: list[DocumentRead]
